src/cast_core/src/RunCommand.cpp

```diff
@@ -208,13 +208,110 @@ class DatabaseOperation {
     bool _awaitStepdown;
 };
 
+// Class to parse and hold the OnlyRunInInstance configuration. This option may be used to only
+// run under certain configurations Available configurations are specified in Options.
+class InstanceTypeFilter {
+public:
+    struct Options {
+        static constexpr std::string_view kStandalone = "standalone";
+        static constexpr std::string_view kSharded = "sharded";
+        static constexpr std::string_view kReplicaSet = "replica_set";
+
+        static std::string getValidOptionsListAsString() {
+            std::stringstream listString;
+            listString << kStandalone << ", " << kSharded << ", " << kReplicaSet;
+            return listString.str();
+        }
+
+        static bool isValid(std::string_view instanceType) {
+            return instanceType == kStandalone || instanceType == kSharded
+                    || instanceType == kReplicaSet;
+        }
+
+        static bool isValidList(const std::vector<std::string>& options) {
+            for(auto &option : options) {
+                if(!isValid(option)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
+
+    InstanceTypeFilter(PhaseContext &context, mongocxx::pool::entry& client) {
+        try {
+            _onlyRunInInstancesContext = context.getPlural<std::string>(
+                "OnlyRunInInstance", "OnlyRunInInstances", [&](const Node& node) {
+                    return node.to<std::string>();
+            });
+        } catch(...) {
+            // Exception might be due to keys not found or other errors. If its due to keys not
+            // found ignore the expception. Otherwise rethrow.
+            if(context["OnlyRunInInstance"] || context["OnlyRunInInstance"]) {
+                std::rethrow_exception(std::current_exception());
+            }
+        }
+
+        if(_onlyRunInInstancesContext.size()) {
+            if(!Options::isValidList(_onlyRunInInstancesContext)) {
+                std::stringstream errorMsg;
+                errorMsg << "OnlyRunInInstance or OnlyRunInInstances valid values are: "
+                    << Options::getValidOptionsListAsString();
+                throw InvalidConfigurationException(errorMsg.str());
+            }
+
+            _adminDB = (*client)["admin"];
+        }
+    }
+
+    // Indicates if operations should be run. If there is no OnlyRunInInstance configuration or one
+    // of the specified options matches the client instance type, returns true. Otherwise returns
+    // false.
+    bool shouldRun() {
+        if(_type.length() == 0 && _onlyRunInInstancesContext.size()) {
+            _type = getInstanceType(_adminDB);
```

Comment:
If shouldRun gets evaluated during the run(), then this `getInstanceType` gets evaluated during the phase rather than during the context setup? We should ideally only run this during context setup, the instance shouldn't change after that point so that's safe.

For some workloads, running this during the phase may be prohibitively expensive.

Reply:
That's what I tried doing in the first place. But it was failing the PR's compilation task while doing a dry-run. I tried running dry locally, and that worked fine. But I wasn't able to find out why it failed in evergreen. See [patch](https://spruce.mongodb.com/version/624f04e53066151c1c109b0d/changes?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) and [error](https://evergreen.mongodb.com/task_log_raw/genny_amazon2_t_compile_patch_25a5f5737671be446e240a71cf6cd28f3829cedd_624f04e53066151c1c109b0d_22_04_07_15_36_06/0?type=T#L899)

The conclusion I reached was that dry mode was supposed to be able to evaluate the configuration, regardless of connection to the database. And that we should not query the database in context setup. So I ended up moving the query to the server out of context setup.

Any idea of why it fails?

Reply:
yeah dry run doesn't run against a server, so makes sense that it'll fail here.

I recommend moving it to the context and putting your workload here to be excluded: https://github.com/mongodb/genny/blob/c38a522e44a12fec43558527f110998eaa817465/src/lamplib/src/genny/tasks/dry_run.py#L28

We have [TIG-3290](https://jira.mongodb.org/browse/TIG-3290) in the future to improve this!

Reply:
In that case every workload using this feature would have to be added to the exclusion list in dry_run.py, wouldn't that be quite confusing and cumbersome? 

The result of hello is cached after the first time. Do you still think that this is prohibitively expensive?

Reply:
I think existing workloads using this actor should see no effect because the configuration for this option will be empty and `shouldRun` will return without doing the query.

But given the concerns for now I will do as suggested, move the query to context and excluding the workload from dry run.

Thanks for all the feedback!